test/plugin/test_out_secondary_file.rb

```diff
@@ -0,0 +1,393 @@
+
+class FileOutputSecondaryTest < Test::Unit::TestCase
+  include Fluent::UniqueId::Mixin
+
+  def setup
+    Fluent::Test.setup
+    FileUtils.rm_rf(TMP_DIR)
+    FileUtils.mkdir_p(TMP_DIR)
+  end
+
+  TMP_DIR = File.expand_path(File.dirname(__FILE__) + "/../tmp/out_secondary_file#{ENV['TEST_ENV_NUMBER']}")
+
+  CONFIG = %[
+    directory #{TMP_DIR}
+    basename out_file_test
+    compress gzip
+  ]
+
+  class DummyOutput < Fluent::Plugin::Output
+    def write(chunk); end
+  end
+
+  def create_primary(buffer_cofig = config_element('buffer'))
+    DummyOutput.new.configure(config_element('ROOT','',{}, [buffer_cofig]))
+  end
+
+  def create_driver(conf = CONFIG, primary = create_primary)
+    c = Fluent::Test::Driver::Output.new(Fluent::Plugin::SecondaryFileOutput)
+    c.instance.acts_as_secondary(primary)
+    c.configure(conf)
+  end
+
+  sub_test_case 'configture' do
+    test 'default configuration' do
+      d = create_driver %[directory #{TMP_DIR}]
+      assert_equal 'dump.bin', d.instance.basename
+      assert_equal TMP_DIR, d.instance.directory
+      assert_equal :text, d.instance.compress
+    end
+
+    test 'should be configurable' do
+      d = create_driver
+      assert_equal 'out_file_test', d.instance.basename
+      assert_equal TMP_DIR, d.instance.directory
+      assert_equal :gzip, d.instance.compress
+    end
+
+    test 'should only use in secondary' do
+      c = Fluent::Test::Driver::Output.new(Fluent::Plugin::SecondaryFileOutput)
+      assert_raise do
+        c.configure(conf)
+      end
+    end
+
+    test 'directory should be writable' do
+      assert_nothing_raised do
+        create_driver %[directory #{TMP_DIR}/test_dir/foo/bar/]
+      end
+
+      assert_nothing_raised do
+        FileUtils.mkdir_p("#{TMP_DIR}/test_dir")
+        File.chmod(0777, "#{TMP_DIR}/test_dir")
+        create_driver %[directory #{TMP_DIR}/test_dir/foo/bar/]
+      end
+
+      assert_raise Fluent::ConfigError do
+        FileUtils.mkdir_p("#{TMP_DIR}/test_dir")
+        File.chmod(0555, "#{TMP_DIR}/test_dir")
+        create_driver %[directory #{TMP_DIR}/test_dir/foo/bar/]
+      end
+    end
+
+    test 'should be passed directory' do
+      assert_raise Fluent::ConfigError do
+        create_driver %[]
+      end
+
+      assert_nothing_raised do
+        create_driver %[directory #{TMP_DIR}/test_dir/foo/bar/]
+      end
+    end
+  end
+
+  def check_gzipped_result(path, expect)
+    # Zlib::GzipReader has a bug of concatenated file: https://bugs.ruby-lang.org/issues/9790
+    # Following code from https://www.ruby-forum.com/topic/971591#979520
+    result = ""
+    File.open(path, "rb") { |io|
+      loop do
+        gzr = Zlib::GzipReader.new(io)
+        result << gzr.read
+        unused = gzr.unused
+        gzr.finish
+        break if unused.nil?
+        io.pos -= unused.length
+      end
+    }
+
+    assert_equal expect, result
+  end
+
+  class DummyMemoryChunk < Fluent::Plugin::Buffer::MemoryChunk; end
```

Comment:
Use `primary.buffer.create_metadata` and `primary.buffer.generate_chunk` instead of methods to simulate these, because it's too dependent to Fluentd's internal code and plugins.
Plugins should NOT depend on internals of other plugins as far as possible, because other plugins are out of control.